### ISSUE TITLE
Ignore padding newlines rule for `function` if keyword is inline

### ISSUE THREAD

**Issue by ashclarke** [jscs-dev/node-jscs#1035]
Hello,

Sometimes it is necessary to return a function, such as:

```
function orderObjectBy() {
    return function (obj, value) {
        // more code
    }
}
```

In this case, if you have `requirePaddingNewlinesBeforeKeywords` set to true, it will throw the error:
`Keyword 'function' should have an empty line above it`. However, sometimes it is desired that this should happen.

Is it possible to ignore this rule if the `function` keyword is inline with `return`, possibly by adding it as an  [excluded token](https://github.com/jscs-dev/node-jscs/commit/0c3887b92e8543220f428ac87b3a49053a564da5)?

This would also apply for functions in arrays, that may have no dependencies, a la Angular:

```
app
    .directive("myDirective", [function () {
        // more code
    }]);
```

**Comment by mikesherov** [comment]
@ashclarke thanks for contributing! We'll look into this shortly.

/cc @markelog @mdevils rules that are keyword driven seem to continually suffer this fate. I've argued against token based rules for a while, but here are the edge cases: `while` can also be `do while`, `default` can now also be `export default`, `function` can be a function expression or a function declaration, `if` can also be an `else if`, which typically has caveats.

What should we do here?

**Comment by jfelchner** [comment]
Any news on this?

**Comment by markelog** [comment]
@jfelchner do you have any suggestion how we could implement such feature?

**Comment by jfelchner** [comment]
@markelog I have zero clue about how all this is set up so sorry, I'm not going to be much help. :(

**Comment by ashclarke** [comment]
In all cases of a "false positive" in my code, the functions are all anonymous. Is this something that could be detected? I assume that would need a hefty configuration change, that would probably be more effort than my next suggestion.

The previous token**?** before each of my function declarations is either the keyword `return` or an array open bracket. Perhaps that is a simpler approach than changing the configuration?

Dodgy pseudocode alert!

    if current keyword is function
      if previous token is return keyword or array open bracket then
        continue
      else
        flag this as bad
      end if
    end if

**Comment by zxqfox** [comment]
I'd say :cat::
```
skip unless current node is function
skip if parent node is return
skip if parent node is array

report this!
```

**Comment by zxqfox** [comment]
@mikesherov @markelog @mdevils Can we remove function from that list and implement a separate rule for functions? I didn't see a good way to not break current solution while fixing this bug.

Also I'm not sure why it was labeled as `enhancement`. It's looks like a bug, and sound like a bug.

**Comment by ashclarke** [comment]
I also have this issue with do while statements, where the `while` keyword is on a new line.

e.g.

```
Keyword `while` should have an empty line above it :
    90 |
    91 |                        node = node.offsetParent;
    92 |                    }
-----------------------------^
    93 |                    while (node);

FROM:

do {
    node = node.offsetParent;
}
while (node);

```

however:
```
thing = [];

while (thing.length) {
    // more code
}
```

or:
```
if (!thing) {
    thing = [];
}

while (thing.length) {
    // more code
}
```

is also valid.

**Comment by hzoo** [comment]
We released [3.0](https://github.com/jscs-dev/node-jscs/releases/tag/v3.0.0) recently! We also wrote a [post](https://medium.com/@markelog/jscs-end-of-the-line-bc9bf0b3fdb2#.oot3wvgpu) about how 3.0 will be our last major release (since we are deprecating the project) and that we are [merging with ESLint](http://eslint.org/blog/2016/04/welcoming-jscs-to-eslint)!

Thus we won't be working on any new features moving forward which includes new rules/options. We will continue to fix bugs and work on CST/better autofixing for current rules in the upcoming months both to help JSCS users and to gain experience that we can hopefully use in ESLint. But otherwise we both teams will be working on getting [ESLint feature parity with JSCS](https://github.com/eslint/eslint/milestones/JSCS%20Compatibility) and just a nice transition for our users.

I'm putting an `orphaned` label to note this issue isn't planned to be worked on, but might be a good contribution as a custom rule (not in JSCS core itself).

I would encourage anyone who wants the rule to try writing it themselves: I would either write a custom rule or submit an issue to ESLint and if accepted a PR. Otherwise a custom rule that won't be in core repo itself.

JSCS rules are also very similar to ESLint rules: here are some resources to get started: 

- [Project-specific Lint Rules with ESLint](https://gist.github.com/jareware/7179093)
- [Creating Custom Rules for ESLint](https://insideops.wordpress.com/2015/12/08/creating-custom-rules-for-eslint)
- [Maintaining JavaScript Code Quality with ESLint ](https://www.paypal-engineering.com/2014/12/12/maintaining-javascript-code-quality-with-eslint/)
